From 47b743cb8c224fa12a73a8786fd89493e031c8af Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sun, 16 Nov 2025 22:02:24 +0100
Subject: [PATCH 15/17] win32u: Introduce a NtUserSetForegroundWindowInternal
 call.

For user drivers to forcefully change foreground window, regardless of
the restriction which will be introduced next.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=58167
---
 dlls/win32u/defwnd.c          | 2 +-
 dlls/win32u/input.c           | 7 ++++---
 dlls/win32u/message.c         | 6 +++---
 dlls/win32u/win32u_private.h  | 2 +-
 dlls/win32u/window.c          | 7 +++++--
 dlls/winemac.drv/window.c     | 8 ++++----
 dlls/winewayland.drv/window.c | 2 +-
 dlls/winex11.drv/event.c      | 6 +++---
 include/ntuser.h              | 6 ++++++
 server/protocol.def           | 1 +
 10 files changed, 29 insertions(+), 18 deletions(-)

diff --git a/dlls/win32u/defwnd.c b/dlls/win32u/defwnd.c
index 4055d5b9a68..1f9c9a4c7cb 100644
--- a/dlls/win32u/defwnd.c
+++ b/dlls/win32u/defwnd.c
@@ -2194,7 +2194,7 @@ static LRESULT handle_nc_lbutton_down( HWND hwnd, WPARAM wparam, LPARAM lparam )
                 top = parent;
             }
 
-            if (set_foreground_window( top, TRUE ) || (get_active_window() == top))
+            if (set_foreground_window( top, TRUE, FALSE ) || (get_active_window() == top))
                 send_message( hwnd, WM_SYSCOMMAND, SC_MOVE + HTCAPTION, lparam );
             break;
         }
diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 3c6249d211e..6e593a2f1bd 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -2138,7 +2138,7 @@ HWND WINAPI NtUserSetFocus( HWND hwnd )
         if (call_hooks( WH_CBT, HCBT_SETFOCUS, (WPARAM)hwnd, (LPARAM)previous, 0 )) return 0;
 
         /* activate hwndTop if needed. */
-        if (!(active = get_active_window()) && !set_foreground_window( hwndTop, FALSE )) return 0;
+        if (!(active = get_active_window()) && !set_foreground_window( hwndTop, FALSE, FALSE )) return 0;
         if (hwndTop != active)
         {
             if (!set_active_window( hwndTop, NULL, FALSE, FALSE, 0 )) return 0;
@@ -2163,13 +2163,13 @@ HWND WINAPI NtUserSetFocus( HWND hwnd )
  */
 BOOL WINAPI NtUserSetForegroundWindow( HWND hwnd )
 {
-    return set_foreground_window( hwnd, FALSE );
+    return set_foreground_window( hwnd, FALSE, FALSE );
 }
 
 /*******************************************************************
  *		set_foreground_window
  */
-BOOL set_foreground_window( HWND hwnd, BOOL mouse )
+BOOL set_foreground_window( HWND hwnd, BOOL mouse, BOOL internal )
 {
     BOOL ret, send_msg_old = FALSE, send_msg_new = FALSE;
     DWORD new_thread_id;
@@ -2181,6 +2181,7 @@ BOOL set_foreground_window( HWND hwnd, BOOL mouse )
     SERVER_START_REQ( set_foreground_window )
     {
         req->handle = wine_server_user_handle( hwnd );
+        req->internal = internal;
         if ((ret = !wine_server_call_err( req )))
         {
             previous = wine_server_ptr_handle( reply->previous );
diff --git a/dlls/win32u/message.c b/dlls/win32u/message.c
index 035ab6574e4..6df5b680cec 100644
--- a/dlls/win32u/message.c
+++ b/dlls/win32u/message.c
@@ -2179,7 +2179,7 @@ static LRESULT handle_internal_message( HWND hwnd, UINT msg, WPARAM wparam, LPAR
         HWND foreground;
 
         if (!user_driver->pGetWindowStateUpdates( hwnd, &state_cmd, &config_cmd, &window_rect, &foreground )) return 0;
-        if (foreground) NtUserSetForegroundWindow( foreground );
+        if (foreground) set_foreground_window( foreground, FALSE, TRUE );
         if (state_cmd)
         {
             if (LOWORD(state_cmd) == SC_RESTORE && HIWORD(state_cmd)) NtUserSetActiveWindow( hwnd );
@@ -2187,7 +2187,7 @@ static LRESULT handle_internal_message( HWND hwnd, UINT msg, WPARAM wparam, LPAR
 
             /* state change might have changed the window config already, check again */
             user_driver->pGetWindowStateUpdates( hwnd, &state_cmd, &config_cmd, &window_rect, &foreground );
-            if (foreground) NtUserSetForegroundWindow( foreground );
+            if (foreground) set_foreground_window( foreground, FALSE, TRUE );
             if (state_cmd) WARN( "window %p state needs another update, ignoring\n", hwnd );
         }
         if (config_cmd)
@@ -2681,7 +2681,7 @@ static BOOL process_mouse_message( MSG *msg, UINT hw_id, ULONG_PTR extra_info, H
                     /* fall through */
                 case MA_ACTIVATE:
                 case 0:
-                    if (!set_foreground_window( hwndTop, TRUE )) eat_msg = TRUE;
+                    if (!set_foreground_window( hwndTop, TRUE, FALSE )) eat_msg = TRUE;
                     break;
                 default:
                     WARN( "unknown WM_MOUSEACTIVATE code %d\n", ret );
diff --git a/dlls/win32u/win32u_private.h b/dlls/win32u/win32u_private.h
index 63e9c517a1f..0506b9c1cd1 100644
--- a/dlls/win32u/win32u_private.h
+++ b/dlls/win32u/win32u_private.h
@@ -95,7 +95,7 @@ extern DWORD get_input_state(void);
 extern DWORD get_last_input_time(void);
 extern BOOL get_async_keyboard_state( BYTE state[256] );
 extern BOOL set_capture_window( HWND hwnd, UINT gui_flags, HWND *prev_ret );
-extern BOOL set_foreground_window( HWND hwnd, BOOL mouse );
+extern BOOL set_foreground_window( HWND hwnd, BOOL mouse, BOOL force );
 extern BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus, DWORD new_active_thread_id );
 extern BOOL set_ime_composition_rect( HWND hwnd, RECT rect );
 extern void toggle_caret( HWND hwnd );
diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index 16373d2e6e9..848e8286298 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -3866,7 +3866,7 @@ BOOL set_window_pos( WINDOWPOS *winpos, int parent_x, int parent_y )
         if ((style & (WS_CHILD | WS_POPUP)) == WS_CHILD)
             send_message( winpos->hwnd, WM_CHILDACTIVATE, 0, 0 );
         else if (!(style & WS_MINIMIZE))
-            set_foreground_window( winpos->hwnd, FALSE );
+            set_foreground_window( winpos->hwnd, FALSE, FALSE );
     }
 
     if(!(orig_flags & SWP_DEFERERASE))
@@ -4211,7 +4211,7 @@ static void activate_other_window( HWND hwnd )
     TRACE( "win = %p fg = %p\n", hwnd_to, fg );
     if (!fg || hwnd == fg)
     {
-        if (set_foreground_window( hwnd_to, FALSE )) return;
+        if (set_foreground_window( hwnd_to, FALSE, FALSE )) return;
     }
     if (NtUserSetActiveWindow( hwnd_to )) NtUserSetActiveWindow( 0 );
 }
@@ -5909,6 +5909,9 @@ ULONG_PTR WINAPI NtUserCallHwnd( HWND hwnd, DWORD code )
         activate_other_window( hwnd );
         return 0;
 
+    case NtUserCallHwnd_SetForegroundWindowInternal:
+        return set_foreground_window( hwnd, FALSE, TRUE );
+
     case NtUserCallHwnd_GetDpiForWindow:
         return get_dpi_for_window( hwnd );
 
diff --git a/dlls/winemac.drv/window.c b/dlls/winemac.drv/window.c
index 765649f40e4..79464944c5d 100644
--- a/dlls/winemac.drv/window.c
+++ b/dlls/winemac.drv/window.c
@@ -1758,7 +1758,7 @@ void macdrv_window_got_focus(HWND hwnd, const macdrv_event *event)
     if (can_window_become_foreground(hwnd) && !(style & WS_MINIMIZE))
     {
         TRACE("setting foreground window to %p\n", hwnd);
-        NtUserSetForegroundWindow(hwnd);
+        NtUserSetForegroundWindowInternal(hwnd);
         return;
     }
 
@@ -1782,7 +1782,7 @@ void macdrv_window_lost_focus(HWND hwnd, const macdrv_event *event)
     {
         send_message(hwnd, WM_CANCELMODE, 0, 0);
         if (hwnd == NtUserGetForegroundWindow())
-            NtUserSetForegroundWindow(NtUserGetDesktopWindow());
+            NtUserSetForegroundWindowInternal(NtUserGetDesktopWindow());
     }
 }
 
@@ -1811,7 +1811,7 @@ void macdrv_app_deactivated(void)
     if (get_active_window() == NtUserGetForegroundWindow())
     {
         TRACE("setting fg to desktop\n");
-        NtUserSetForegroundWindow(NtUserGetDesktopWindow());
+        NtUserSetForegroundWindowInternal(NtUserGetDesktopWindow());
     }
 }
 
@@ -1975,7 +1975,7 @@ void macdrv_window_drag_begin(HWND hwnd, const macdrv_event *event)
         if (ma != MA_NOACTIVATEANDEAT && ma != MA_NOACTIVATE)
         {
             TRACE("setting foreground window to %p\n", hwnd);
-            NtUserSetForegroundWindow(hwnd);
+            NtUserSetForegroundWindowInternal(hwnd);
         }
     }
 
diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index 79a6a584728..b088c3337eb 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -654,7 +654,7 @@ LRESULT WAYLAND_WindowMessage(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp)
         wayland_configure_window(hwnd);
         return 0;
     case WM_WAYLAND_SET_FOREGROUND:
-        NtUserSetForegroundWindow(hwnd);
+        NtUserSetForegroundWindowInternal(hwnd);
         return 0;
     default:
         FIXME("got window msg %x hwnd %p wp %lx lp %lx\n", msg, hwnd, (long)wp, lp);
diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index f4715cd51f7..8e896e4ba1d 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -665,7 +665,7 @@ static void set_focus( Display *display, HWND focus, Time time )
 
     if (!is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ))
     {
-        NtUserSetForegroundWindow( focus );
+        NtUserSetForegroundWindowInternal( focus );
 
         threadinfo.cbSize = sizeof(threadinfo);
         NtUserGetGUIThreadInfo( 0, &threadinfo );
@@ -887,7 +887,7 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
         if (!hwnd) hwnd = x11drv_thread_data()->last_focus;
         if (hwnd && can_activate_window(hwnd)) set_focus( event->display, hwnd, CurrentTime );
     }
-    else NtUserSetForegroundWindow( hwnd );
+    else NtUserSetForegroundWindowInternal( hwnd );
     return TRUE;
 }
 
@@ -918,7 +918,7 @@ static void focus_out( Display *display , HWND hwnd )
         if (hwnd == NtUserGetForegroundWindow())
         {
             TRACE( "lost focus, setting fg to desktop\n" );
-            NtUserSetForegroundWindow( NtUserGetDesktopWindow() );
+            NtUserSetForegroundWindowInternal( NtUserGetDesktopWindow() );
         }
     }
  }
diff --git a/include/ntuser.h b/include/ntuser.h
index 029a8bea8f2..5d4899f0b12 100644
--- a/include/ntuser.h
+++ b/include/ntuser.h
@@ -1293,6 +1293,7 @@ enum
     NtUserCallHwnd_IsWindowEnabled,
     NtUserCallHwnd_IsWindowUnicode,
     NtUserCallHwnd_IsWindowVisible,
+    NtUserCallHwnd_SetForegroundWindowInternal,
     /* temporary exports */
     NtUserGetFullWindowHandle,
     NtUserIsCurrentProcessWindow,
@@ -1304,6 +1305,11 @@ static inline void NtUserActivateOtherWindow( HWND hwnd )
     NtUserCallHwnd( hwnd, NtUserCallHwnd_ActivateOtherWindow );
 }
 
+static inline BOOL NtUserSetForegroundWindowInternal( HWND hwnd )
+{
+    return NtUserCallHwnd( hwnd, NtUserCallHwnd_SetForegroundWindowInternal );
+}
+
 static inline void *NtUserGetDialogInfo( HWND hwnd )
 {
     return (void *)NtUserCallHwnd( hwnd, NtUserCallHwnd_GetDialogInfo );
diff --git a/server/protocol.def b/server/protocol.def
index 5eee7d35a34..9e221b79f8b 100644
--- a/server/protocol.def
+++ b/server/protocol.def
@@ -3107,6 +3107,7 @@ enum coords_relative
 /* Set the system foreground window */
 @REQ(set_foreground_window)
     user_handle_t  handle;        /* handle to the foreground window */
+    int            internal;      /* foreground change comes from the host */
 @REPLY
     user_handle_t  previous;      /* handle to the previous foreground window */
     int            send_msg_old;  /* whether we have to send a msg to the old window */
-- 
2.51.2

